scripts/toltec/bash.py

```diff
@@ -0,0 +1,379 @@
+    "TERM",
+    "UID",
+    "_",
+}
+
+
+def get_declarations(src: str) -> Tuple[Variables, Functions]:
+    """
+    Extract all variables and functions defined by a Bash script.
+
+    If a function or a variable is defined or assigned multiple times
+    in the script, only the final value is extracted. The script must not
+    output anything on the standard output stream.
+
+    :param src: source string of the considered Bash string
+    :returns: a tuple containing the declared variables and functions
+    """
+    src += """
+declare -f
+declare -p
+"""
+    env: Dict[str, str] = {}
+
+    declarations_subshell = (
+        subprocess.run(  # pylint:disable=subprocess-run-check
+            ["/usr/bin/env", "bash"],
+            input=src.encode(),
+            capture_output=True,
+            env=env,
+        )
+    )
+
+    if declarations_subshell.returncode == 2:
+        raise ScriptError(
+            f"Bash syntax error\n\
+{declarations_subshell.stderr.decode()}"
+        )
+
+    if declarations_subshell.returncode != 0:
+        raise ScriptError(
+            f"Bash error\n\
+{declarations_subshell.stderr.decode()}"
+        )
+
+    declarations = declarations_subshell.stdout.decode()
+
+    # Parse `declare` statements and function statements
+    lexer = shlex.shlex(declarations, posix=True)
+    lexer.wordchars = lexer.wordchars + "-"
+
+    variables = {}
+    functions = {}
+
+    while True:
+        token = lexer.get_token()
+
+        if token == lexer.eof:
+            break
+
+        next_token = lexer.get_token()
+
+        if token == "declare" and next_token[0] == "-":
+            lexer.push_token(next_token)
+            name, value = _parse_var(lexer)
+
+            if name not in default_variables:
+                variables[name] = value
+        else:
+            assert next_token == "("
+            assert lexer.get_token() == ")"
+            start, end = _parse_func(lexer)
+            functions[token] = declarations[start:end]
+
+    return variables, functions
+
+
+def put_variables(variables: Variables) -> str:
+    """
+    Generate a Bash script fragment which defines a set of variables.
+
+    :param variables: set of variables to define
+    :returns: generated Bash fragment
+    """
+    result = ""
+
+    for name, value in variables.items():
+        if value is None:
+            result += f"declare -- {name}\n"
+        elif isinstance(value, str):
+            result += f"declare -- {name}={_generate_string(value)}\n"
+        elif isinstance(value, list):
+            result += f"declare -a {name}={_generate_indexed(value)}\n"
+        elif isinstance(value, dict):
+            result += f"declare -A {name}={_generate_assoc(value)}\n"
+        else:
+            raise ValueError(
+                f"Unsupported type {type(value)} for variable \
+{name}"
+            )
+
+    return result
+
+
+def _parse_string(token: str) -> str:
+    """Remove escape sequences from a Bash string."""
```

Comment:
is this the only escape sequence that needs to be removed?

Reply:
As far as I know, yes. The lexer already removes all escape sequences. But the Bash `declare` builtin seems to add an additional escape in front of dollar signs in strings, even where it’s not necessary. Unfortunately this is not documented anywhere I could find.